userspace: compartmentalized app memory organization

cmSAMIAMnot opened the pull request:
Summary: revised attempt at addressing issue 6290.  The
following provides an alternative to using
CONFIG_APPLICATION_MEMORY by compartmentalizing data into
Memory Domains.  Dependent on MPU limitations, supports
compartmentalized Memory Domains for 1...N logical
applications.  This is considered an initial attempt at
designing flexible compartmentalized Memory Domains for
multiple logical applications and, with the provided python
script and edited CMakeLists.txt, provides support for power
of 2 aligned MPU architectures.

Overview: The current patch uses qualifiers to group data into
subsections.  The qualifier usage allows for dynamic subsection
creation and affords the developer a large amount of flexibility
in the grouping, naming, and size of the resulting partitions and
domains that are built on these subsections. By additional macro
calls, functions are created that help calculate the size,
address, and permissions for the subsections and enable the
developer to control application data in specified partitions and
memory domains.

Background: Initial attempts focused on creating a single
section in the linker script that then contained internally
grouped variables/data to allow MPU/MMU alignment and protection.
This did not provide additional functionality beyond
CONFIG_APPLICATION_MEMORY as we were unable to reliably group
data or determine their grouping via exported linker symbols.
Thus, the resulting decision was made to dynamically create
subsections using the current qualifier method. An attempt to
group the data by object file was tested, but found that this
broke applications such as ztest where two object files are
created: ztest and main.  This also creates an issue of grouping
the two object files together in the same memory domain while
also allowing for compartmenting other data among threads.

Because it is not possible to know a) the name of the partition
and thus the symbol in the linker, b) the size of all the data
in the subsection, nor c) the overall number of partitions
created by the developer, it was not feasible to align the
subsections at compile time without using dynamically generated
linker script for MPU architectures requiring power of 2
alignment.

In order to provide support for MPU architectures that require a
power of 2 alignment, a python script is run at build prior to
when linker_priv_stacks.cmd is generated.  This script scans the
built object files for all possible partitions and the names given
to them. It then generates a linker file (app_smem.ld) that is
included in the main linker.ld file.  This app_smem.ld allows the
compiler and linker to then create each subsection and align to
the next power of 2.

Usage:
 - Requires: app_memory/app_memdomain.h .
 - _app_dmem(id) marks a variable to be placed into a data
section for memory partition id.
 - _app_bmem(id) marks a variable to be placed into a bss
section for memory partition id.
 - These are seen in the linker.map as "data_smem_id" and
"data_smem_idb".
 - To create a k_mem_partition, call the macro
app_mem_partition(part0) where "part0" is the name then used to
refer to that partition. This macro only creates a function and
necessary data structures for the later "initialization".
 - To create a memory domain for the partition, the macro
app_mem_domain(dom0) is called where "dom0" is the name then
used for the memory domain.
 - To initialize the partition (effectively adding the partition
to a linked list), init_part_part0() is called. This is followed
by init_app_memory(), which walks all partitions in the linked
list and calculates the sizes for each partition.
 - Once the partition is initialized, the domain can be
initialized with init_domain_dom0(part0) which initializes the
domain with partition part0.
 - After the domain has been initialized, the current thread
can be added using add_thread_dom0(k_current_get()).
 - The code used in ztests ans kernel/init has been added under
a conditional #ifdef to isolate the code from other tests.
The userspace test CMakeLists.txt file has commands to insert
the CONFIG_APP_SHARED_MEM definition into the required build
targets.
  Example:
        /* create partition at top of file outside functions */
        app_mem_partition(part0);
        /* create domain */
        app_mem_domain(dom0);
        _app_dmem(dom0) int var1;
        _app_bmem(dom0) static volatile int var2;

        int main()
        {
                init_part_part0();
                init_app_memory();
                init_domain_dom0(part0);
                add_thread_dom0(k_current_get());
                ...
        }

 - If multiple partitions are being created, a variadic
preprocessor macro can be used as provided in
app_macro_support.h:

        FOR_EACH(app_mem_partition, part0, part1, part2);

or, for multiple domains, similarly:

        FOR_EACH(app_mem_domain, dom0, dom1);

Similarly, the init_part_* can also be used in the macro:

        FOR_EACH(init_part, part0, part1, part2);

Testing:
 - This has been successfully tested on qemu_x86 and the
ARM frdm_k64f board.  It compiles and builds power of 2
aligned subsections for the linker script on the 96b_carbon
boards.  These power of 2 alignments have been checked by
hand and are viewable in the zephyr.map file that is
produced during build. However, due to a shortage of
available MPU regions on the 96b_carbon board, we are unable
to test this.
 - When run on the 96b_carbon board, the test suite will
enter execution, but each individaul test will fail due to
an MPU FAULT.  This is expected as the required number of
MPU regions exceeds the number allowed due to the static
allocation. As the MPU driver does not detect this issue,
the fault occurs because the data being accessed has been
placed outside the active MPU region.
 - This now compiles successfully for the ARC boards
em_starterkit_em7d and em_starterkit_em7d_v22. However,
as we lack ARC hardware to run this build on, we are unable
to test this build.

Current known issues:
1) While the script and edited CMakeLists.txt creates the
ability to align to the next power of 2, this does not
address the shortage of available MPU regions on certain
devices (e.g. 96b_carbon).  In testing the APB and PPB
regions were commented out.
2) checkpatch.pl lists several issues regarding the
following:
a) Complex macros. The FOR_EACH macros as defined in
app_macro_support.h are listed as complex macros needing
parentheses.  Adding parentheses breaks their
functionality, and we have otherwise been unable to
resolve the reported error.
b) __aligned() preferred. The _app_dmem_pad() and
_app_bmem_pad() macros give warnings that __aligned()
is preferred. Prior iterations had this implementation,
which resulted in errors due to "complex macros".
c) Trailing semicolon. The macro init_part(name) has
a trailing semicolon as the semicolon is needed for the
inlined macro call that is generated when this macro
expands.

The previous pull #6827 has been passed to me to follow up.

Signed-off-by: Joshua Domagalski <jedomag@tycho.nsa.gov>

Signed-off-by: Shawn Mosley <smmosle@tycho.nsa.gov>

cmSAMIAMnot committed:
userspace: compartmentalized app memory organization

Summary: revised attempt at addressing issue 6290.  The
following provides an alternative to using
CONFIG_APPLICATION_MEMORY by compartmentalizing data into
Memory Domains.  Dependent on MPU limitations, supports
compartmentalized Memory Domains for 1...N logical
applications.  This is considered an initial attempt at
designing flexible compartmentalized Memory Domains for
multiple logical applications and, with the provided python
script and edited CMakeLists.txt, provides support for power
of 2 aligned MPU architectures.

Overview: The current patch uses qualifiers to group data into
subsections.  The qualifier usage allows for dynamic subsection
creation and affords the developer a large amount of flexibility
in the grouping, naming, and size of the resulting partitions and
domains that are built on these subsections. By additional macro
calls, functions are created that help calculate the size,
address, and permissions for the subsections and enable the
developer to control application data in specified partitions and
memory domains.

Background: Initial attempts focused on creating a single
section in the linker script that then contained internally
grouped variables/data to allow MPU/MMU alignment and protection.
This did not provide additional functionality beyond
CONFIG_APPLICATION_MEMORY as we were unable to reliably group
data or determine their grouping via exported linker symbols.
Thus, the resulting decision was made to dynamically create
subsections using the current qualifier method. An attempt to
group the data by object file was tested, but found that this
broke applications such as ztest where two object files are
created: ztest and main.  This also creates an issue of grouping
the two object files together in the same memory domain while
also allowing for compartmenting other data among threads.

Because it is not possible to know a) the name of the partition
and thus the symbol in the linker, b) the size of all the data
in the subsection, nor c) the overall number of partitions
created by the developer, it was not feasible to align the
subsections at compile time without using dynamically generated
linker script for MPU architectures requiring power of 2
alignment.

In order to provide support for MPU architectures that require a
power of 2 alignment, a python script is run at build prior to
when linker_priv_stacks.cmd is generated.  This script scans the
built object files for all possible partitions and the names given
to them. It then generates a linker file (app_smem.ld) that is
included in the main linker.ld file.  This app_smem.ld allows the
compiler and linker to then create each subsection and align to
the next power of 2.

Usage:
 - Requires: app_memory/app_memdomain.h .
 - _app_dmem(id) marks a variable to be placed into a data
section for memory partition id.
 - _app_bmem(id) marks a variable to be placed into a bss
section for memory partition id.
 - These are seen in the linker.map as "data_smem_id" and
"data_smem_idb".
 - To create a k_mem_partition, call the macro
app_mem_partition(part0) where "part0" is the name then used to
refer to that partition. This macro only creates a function and
necessary data structures for the later "initialization".
 - To create a memory domain for the partition, the macro
app_mem_domain(dom0) is called where "dom0" is the name then
used for the memory domain.
 - To initialize the partition (effectively adding the partition
to a linked list), init_part_part0() is called. This is followed
by init_app_memory(), which walks all partitions in the linked
list and calculates the sizes for each partition.
 - Once the partition is initialized, the domain can be
initialized with init_domain_dom0(part0) which initializes the
domain with partition part0.
 - After the domain has been initialized, the current thread
can be added using add_thread_dom0(k_current_get()).
 - The code used in ztests ans kernel/init has been added under
a conditional #ifdef to isolate the code from other tests.
The userspace test CMakeLists.txt file has commands to insert
the CONFIG_APP_SHARED_MEM definition into the required build
targets.
  Example:
        /* create partition at top of file outside functions */
        app_mem_partition(part0);
        /* create domain */
        app_mem_domain(dom0);
        _app_dmem(dom0) int var1;
        _app_bmem(dom0) static volatile int var2;

        int main()
        {
                init_part_part0();
                init_app_memory();
                init_domain_dom0(part0);
                add_thread_dom0(k_current_get());
                ...
        }

 - If multiple partitions are being created, a variadic
preprocessor macro can be used as provided in
app_macro_support.h:

        FOR_EACH(app_mem_partition, part0, part1, part2);

or, for multiple domains, similarly:

        FOR_EACH(app_mem_domain, dom0, dom1);

Similarly, the init_part_* can also be used in the macro:

        FOR_EACH(init_part, part0, part1, part2);

Testing:
 - This has been successfully tested on qemu_x86 and the
ARM frdm_k64f board.  It compiles and builds power of 2
aligned subsections for the linker script on the 96b_carbon
boards.  These power of 2 alignments have been checked by
hand and are viewable in the zephyr.map file that is
produced during build. However, due to a shortage of
available MPU regions on the 96b_carbon board, we are unable
to test this.
 - When run on the 96b_carbon board, the test suite will
enter execution, but each individaul test will fail due to
an MPU FAULT.  This is expected as the required number of
MPU regions exceeds the number allowed due to the static
allocation. As the MPU driver does not detect this issue,
the fault occurs because the data being accessed has been
placed outside the active MPU region.
 - This now compiles successfully for the ARC boards
em_starterkit_em7d and em_starterkit_em7d_v22. However,
as we lack ARC hardware to run this build on, we are unable
to test this build.

Current known issues:
1) While the script and edited CMakeLists.txt creates the
ability to align to the next power of 2, this does not
address the shortage of available MPU regions on certain
devices (e.g. 96b_carbon).  In testing the APB and PPB
regions were commented out.
2) checkpatch.pl lists several issues regarding the
following:
a) Complex macros. The FOR_EACH macros as defined in
app_macro_support.h are listed as complex macros needing
parentheses.  Adding parentheses breaks their
functionality, and we have otherwise been unable to
resolve the reported error.
b) __aligned() preferred. The _app_dmem_pad() and
_app_bmem_pad() macros give warnings that __aligned()
is preferred. Prior iterations had this implementation,
which resulted in errors due to "complex macros".
c) Trailing semicolon. The macro init_part(name) has
a trailing semicolon as the semicolon is needed for the
inlined macro call that is generated when this macro
expands.

Update: updated to alternative CONFIG_APPLCATION_MEMORY.
Added config option CONFIG_APP_SHARED_MEM to enable a new section
app_smem to contain the shared memory component.  This commit
seperates the Kconfig definition from the definition used for the
conditional code.  The change is in response to changes in the
way the build system treats definitions.  The python script used
to generate a linker script for app_smem was also midified to
simplify the alignment directives.  A default linker script
app_smem.ld was added to remove the conditional includes dependency
on CONFIG_APP_SHARED_MEM.  By addining the default linker script
the prebuild stages link properly prior to the python script running

Signed-off-by: Joshua Domagalski <jedomag@tycho.nsa.gov>
Signed-off-by: Shawn Mosley <smmosle@tycho.nsa.gov>

andrewboie commented:
HI guys,

Will be studying this week, one thing that sticks out that there's no documentation for this new feature.
Would it be possible to take what you have in the commit message, and create a documentation page for this feature under doc/kernel/usermode?

More feedback to come later this week.

andrewboie commented:
BTW don't worry about the checkpatch "complex macros" errors. They have now been downgraded to warnings as they tended to occur anytime someone was doing something nontrivial with the preprocessor.

carlescufi commented:
@cmSAMIAMnot can you please rebase this against the latest master so that only your commits show for review?

carlescufi commented:
@cmSAMIAMnot I was wondering if there was something strange because I tried rebasing your branch and Git had a hard time, but it looks like it actually managed. Check out https://github.com/carlescufi/zephyr/tree/app_shared_smem for a rebased version.

cmSAMIAMnot commented:
The branch should be ok now.  I am looking at a problem with a test that is now failing because app_bss is writable in the userspace threads.  The problem was introduced when I rebased to master last week.  I suspect app_bss is enabled during the initial setup. I have been looking at qemu_x86 and think the page table may enable it when it initially loads.  I would appreciate suggestion for other places to look if anyone is familiar with the boot process relating to app_bss.

carlescufi commented:
CC @pizi-nordic @ioannisg @agross-linaro

cmSAMIAMnot commented:
I started the conversion of the commit message to a document.  I am looking at creating a sample userspace application.  I will start a separate pull for the sample and add the documentation with the sample.

carlescufi commented:
@andrewboie @andyross @agross-linaro Any comments on this one?

andrewboie commented:
@carlescufi @cmSAMIAMnot  if you grep for "userspace" in testcase.yaml you can find all the tests which run threads in user mode.

About the issue you are having, I'll try to take a look this week

cmSAMIAMnot commented:
I fixed a merge conflict and pushed the update

cmSAMIAMnot commented:
checks all failing because of a change to "application_development" in the main branch.  None of the failed tests are connected with my pull

carlescufi commented:
@cmSAMIAMnot 
>checks all failing because of a change to "application_development" in the main branch.

Not sure I understand, the checks are passing in other Pull Requests, what exactly is the issue?

cmSAMIAMnot commented:
Shippable Run 18046 status is FAILED. When I checked the tests application_development test project was the problem.
Test 3
/application_development/cpp/test'
[ 91%] Built target arch__arm__core
In file included from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/kernel/include/kernel_structs.h:180:0,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/app_memory/app_memdomain.h:7,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/ztest/include/ztest_test.h:16,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/ztest/include/ztest.h:59,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/application_development/cpp/src/main.cpp:32:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/arm/include/kernel_arch_func.h: In function ‘void _arch_switch_to_main_thread(k_thread*, k_thread_stack_t*, size_t, k_thread_entry_t)’:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/arm/include/kernel_arch_func.h:56:24: error: invalid conversion from ‘void*’ to ‘char*’ [-fpermissive]
  start_of_main_stack = (void *)STACK_ROUND_DOWN(start_of_main_stack);
Test 4
/application_development/cpp/test'
[ 33%] Building C object zephyr/CMakeFiles/zephyr.dir/arch/common/isr_tables.c.obj
[ 34%] Building ASM object zephyr/CMakeFiles/zephyr.dir/arch/riscv32/soc/riscv-privilege/common/soc_irq.S.obj
[ 35%] Building C object zephyr/CMakeFiles/zephyr.dir/arch/riscv32/core/cpu_idle.c.obj
[ 36%] Building ASM object zephyr/CMakeFiles/zephyr.dir/arch/riscv32/soc/riscv-privilege/riscv32-qemu/vector.S.obj
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/application_development/cpp/src/main.cpp:55:1: error: expected ‘}’ at end of input
 }

Test 5
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/xtensa/include/kernel_arch_func.h:39:9: error: invalid conversion from ‘void*’ to ‘_cpu_t* {aka _cpu*}’ [-fpermissive]
  return val

I do not have any code touching those files.  And when I performed a fetch/rebase none of them were reported as modified, so I assume it must be the base.

cmSAMIAMnot commented:
I changed app_memdomain.h by adding "#include <kernel_structs.h>" since it is a required dependency.
and I changed include/misc/util.h due to a merge conflict.  In util.h I moved my additions to the new end of file.

carlescufi commented:
@cmSAMIAMnot I can reproduce the failure locally.

Could it be that you are missing:

```
#ifdef __cplusplus
extern "C" {
#endif
```

in a new header file that you have introduced? since the test that fails is the C++ build test

cmSAMIAMnot commented:
when I remove the include the error goes away.  I can make the change and just document the need to add "#include <kernel_structs.h>" when creating a project using this code

carlescufi commented:
>when I remove the include the error goes away. I can make the change and just document the need to add "#include <kernel_structs.h>" when creating a project using this code

Have you tried adding the c++ guards to your .h file?

cmSAMIAMnot commented:
no I have not.  it probably wound not be a bad idea.  The include file does not seem to have a problem when when kernel_structs.h is missing.  I will look at the C++ issues when I create a sample program using C++.

cmSAMIAMnot commented:
Has anyone looked at adding an MPU exception handler to cover cases where there are more partitions assigned to a memory domain than the hardware supports?  The basic idea would be a exception occurs and the handler checks the current thread's memory domain table for the range, and if it is found one of the regions in the MPU would be set to cover that range.  The thread could then be reset to continue running.  The operation would cause an obvious penalty for changing the partitions, but the application developer would not be limited by the number of regions in the MPU.  For simplicity, the partitions could be ordered by the developer based on the entry to the domain, where the first few partitions are never swapped.  If something like this was implemented, applications could could use the app_memory subsystem without needing to tailor the number of partitions to the hardware platform.

stephensmalley commented:
Sorry, I tried to resolve a conflict via the GitHub interface since cmSAMIAMnot is away this week and it created a merge commit rather than squashing and re-basing.  Will have to have him fix it when he returns.  I have tested this commit on qemu_x86, frdm_k64f, and 96b_carbon successfully.  Are there any residual objections/concerns with it?  He has some further work which builds upon it in a sample application to demonstrate isolation, but it depends on this commit.

stephensmalley reviewed (approved):
Reviewed and tested on qemu_x86, frdm_k64f (NXP MPU), and 96b_carbon (ARM MPU).

andrewboie commented:
@stephensmalley I'm eager and willing to take this if we can do the following:

1) Get CI passing all tests except for nonsense checkpatch.pl errors
2) Add complete documentation for this feature under doc/

The documentation is important and will also help with code review. We need to move the instructions out of the commit message and under doc/.

Once this is in, we have about a dozen or so test cases and samples which use user mode threads. I would like to convert them all to use this mechanism instead, and completely remove CONFIG_APPLICATION_MEMORY from the tree for 1.13. We can help out with this.

carlescufi commented:
>Will have to have him fix it when he returns

@stephensmalley when one creates a fork on GitHub, by default all branches are public and writeable by anyone. You should be able to take @cmSAMIAMnot's branch and push to it without any problems.

andrewboie commented:
One question: how do we deal with library code that uses global variables? Do we need to tag all of these as going into some common shared section?

For example some of the heap related stuff in lib/libc/newlib/libc-hooks.c.

cmSAMIAMnot commented:
I have thought about three options.  Create a definition for the globals to assign them to a memory domain.   Add a process prior to linking to perform a section rename of the lib globals to a memory domain section.  The last would be to allow globals as a static region and include a test for user code to cause an error when globals are detected.
I like the first option as a long term solution.  The other two options cause less impact on the code, but they slightly complicate the build process.

stephensmalley commented on scripts/sanitycheck:
```diff
@@ -662,7 +662,7 @@ class SizeCalculator:
                    "ccm_data", "usb_descriptor", "usb_data", "usb_bos_desc",
                    'log_backends_sections', 'log_dynamic_sections',
-                   'log_const_sections']
+                   'log_const_sections',"app_smem"]
```
I noticed that this is now using two different quoting styles (" vs '), not sure why (you didn't introduce it and were using the original style but subsequent commits have switched to single quotes) or what's preferred.

stephensmalley commented on subsys/app_memory/Kconfig:
```diff
@@ -0,0 +1,9 @@
+config APP_SHARED_MEM
+	bool
+	prompt "alt application shared memory"
+	default n
```
We obviously will want a better description for this option ultimately, maybe "Support fine-grained application memory sharing" or something.

stephensmalley commented on subsys/app_memory/Kconfig:
```diff
@@ -0,0 +1,9 @@
+	prompt "alt application shared memory"
+	default n
+	help
+	 This is a testing option
```
Likewise, we'll want better help text.  Maybe "Enable support for selective sharing of global variables among application threads" and include a reference to the doc file?

stephensmalley reviewed (approved):
Looks good aside from a few minor fixes.  Passed userspace tests for me on qemu_x86, frdm_k64f, and 96b_carbon.

stephensmalley commented:
Can we get another review (or if there are no further comments, a merge) of this PR?

andrewboie commented:
I want to have a couple more people look at this in detail but this LGTM.

dcpleung commented on arch/x86/core/x86_mmu.c:
```diff
@@ -25,6 +25,8 @@ MMU_BOOT_REGION((u32_t)&_image_rom_start, (u32_t)&_image_rom_size,
 		MMU_ENTRY_WRITE | MMU_ENTRY_USER | MMU_ENTRY_EXECUTE_DISABLE);
 #endif
+MMU_BOOT_REGION((u32_t)&_app_smem_start, (u32_t)&_app_smem_size,
+		MMU_ENTRY_WRITE | MMU_ENTRY_USER | MMU_ENTRY_EXECUTE_DISABLE);
```
Does this need to be wrapped under CONFIG_APP_SHARED_MEM?

AdithyaBaglody replied:
Agreed, this has to be inside a ifdef block.

AdithyaBaglody commented on CMakeLists.txt:
```diff
@@ -1245,3 +1273,14 @@ if(CONFIG_X86 AND CONFIG_USERSPACE AND NOT CONFIG_X86_NO_MELTDOWN)
 )
 endif()
+
+if(CONFIG_APP_SHARED_MEM)
```
Where exactly are you facing the linker issue in prebuilt? 
As of now I am able to build and run without any issues when I disable this section.

AdithyaBaglody commented on tests/kernel/mem_protect/userspace/CMakeLists.txt:
```diff
@@ -3,3 +3,9 @@ project(NONE)
 target_sources(app PRIVATE ${app_sources})
+
+
+#add_definitions( -DCONFIG_APP_SHARED_MEM )
```
I think these are not needed here.

AdithyaBaglody commented on tests/ztest/include/ztest_test.h:
```diff
@@ -139,8 +141,17 @@ static inline void unit_test_noop(void)
  */
+
+/* definitions for use with testing application shared memory   */
+#ifdef APP_SHARED_MEM
```
Maybe CONFIG_APP_SHARED_MEM can be used instead of this. I feel its easier to maintain.

AdithyaBaglody commented on subsys/CMakeLists.txt:
```diff
@@ -1,3 +1,4 @@
+add_subdirectory(app_memory)
```
I guess we can make this into add_subdirectory_ifdef(CONFIG_APP_SHARED_MEM, app_memory)

cmSAMIAMnot commented:
i implemented all the suggested changes, and added "extra_sections: app_smem" to the testcase.yaml file.  It looks like the sanitycheck script changed.  Even with the section name in the rw_sections, the script reported a error on the userspace tests related to the app_smem section.  The update to testcase.yaml fixed the error.

andrewboie commented on tests/kernel/mem_protect/userspace/testcase.yaml:
```diff
@@ -2,3 +2,4 @@ tests:
   kernel.memory_protection.userspace:
     filter: CONFIG_ARCH_HAS_USERSPACE
     tags: core security userspace ignore_faults
+    extra_sections: app_smem
```
It's going to be cumbersome to have to update more and more YAML as tests are converted to use this.
Let's try to root-cause what's going on with sanitycheck,

cmSAMIAMnot replied:
So, I inaccurately assumed it was sanity check.  Something in the way i have my directory structure currently causes the zephyr-env.sh to set the path for sanitycheck to a different instance of zephyr in a parent directory that did not have the modifications.  Thus requiring the modification to the yaml.

andrewboie commented:
Please also fix these checkpatch issues

```
-:76: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#76: FILE: arch/x86/core/x86_mmu.c:28:
+#ifdef ^ICONFIG_APP_SHARED_MEM$

-:76: ERROR:SPACING: exactly one space required after that #ifdef
#76: FILE: arch/x86/core/x86_mmu.c:28:
+#ifdef 	CONFIG_APP_SHARED_MEM

-:237: WARNING:PREFER_ALIGNED: __aligned(size) is preferred over __attribute__((aligned(size)))
#237: FILE: include/app_memory/app_memdomain.h:30:
+	__attribute__((aligned(MEM_DOMAIN_ALIGN_SIZE), \

-:241: WARNING:PREFER_ALIGNED: __aligned(size) is preferred over __attribute__((aligned(size)))
#241: FILE: include/app_memory/app_memdomain.h:34:
+	__attribute__((aligned(MEM_DOMAIN_ALIGN_SIZE), \

-:330: WARNING:TRAILING_SEMICOLON: macros should not use a trailing semicolon
#330: FILE: include/app_memory/app_memdomain.h:123:
+#define appmem_init_part(name) \
+	appmem_init_part_##name();
```
